vignettes/data-repository.Rmd

```diff
@@ -0,0 +1,200 @@
+
+Git stores the version history under the form of diffs: a list of lines which are deleted and a list of lines which are inserted at a specific line number in a file. We have to keep this in mind when storing data as plain text files in Git. The table below indicate what impact changing the information content has on the diff in Git. Note that changing observations requires a small diff, while changing variables results in a massive diff.
+
+```{r echo = FALSE}
+cases <- rbind(
+    c("remove 1 observation", "remove 1 line"),
+    c("add 1 observation", "add 1 line"),
+    c("update 1 observation", "remove 1 line and add 1 line"),
+    c("remove 1 variable", "remove all lines and add all lines"),
+    c("add 1 variable", "remove all lines and add all lines")
+)
+colnames(cases) <- c("Change in data", "Git diff")
+kable(cases, caption = "Changes in the information content of the data")
+```
+
+In the next table we show the effect on the diff when making changes in the data which don't change in the information content in the data.
+
+```{r echo = FALSE}
+cases <- rbind(
+    c("move 1 observation", "remove 1 line and add 1 line"),
+    c("move 1 variable", "remove all lines and add all lines")
+)
+colnames(cases) <- c("Change in data", "Git diff")
+kable(cases, caption = "Changes in the data without changing the information content")
+```
+
+So in order to use Git a a performant and efficient version control system for data, we need to make sure to store the meta data and keep the diffs as small as possible.
+
+## Tools in `git2r`
+
+### Data repository
+
+A 'data repository' refers to a sub directory within the root of an existing git repository. Create a `data_repository` object with `init()` or `repository()` and specify the sub directory through the `project` argument. Most functions of the package handle a `data_repository` object in the same way as they handle a `git_repository` object. The main difference is that the working directory points to the sub directory of the root, rather that the root of the git repository. You can create multiple data repositories within the same git repository.
+
+```{r}
+# Create a directory in tempdir
+path <- tempfile(pattern = "git2r-")
+dir.create(path)
+
+library(git2r)
+# inititialize a new git repository
+repo <- init(path)
+# use a data repository in an existing git repository
+data_repo <- repository(path, project = "test")
+```
+
+```{r eval = FALSE}
+# inititialize a new git repository and use a data repository within it
+data_repo <- init(path, project = "test")
+```
+
+### Storing data
+
+Use `write_delim_git()` to store a `data.frame` into the data repository. The function will separate the data and the meta data. The data is stored as a headerless, unquoted tab delimited file with ".tsv" extension and UTF-8 encoding. The meta data is stored in YAML format with ".yml" format. Therefore any extension given to the `file` will be stripped (with a warning).
```

Comment:
cfr. also here, a repository to me is a git-repository (there's only one repo for a project) and I do not want to think about (nested) data-repositories inside a git-repository...  So, I would rather have *Use `write_delim_git()` to store a `data.frame` into the data repository.* -> *Use `write_delim_git()` to store a `data.frame` into the (git) repository.*

Reply:
I do understand the reasoning, but you actually loose some interesting Github features (online checking and filtering) when storing headerless + having factors/posix saved more efficiently...

Reply:
Can you point me to an example of this online checking and filtering?

Reply:
For example, [this datafile](https://github.com/trias-project/uredinales-belgium-checklist/blob/master/data/processed/distribution.csv), the *search this file* dialogue:
![image](https://user-images.githubusercontent.com/754862/44341222-8db83500-a487-11e8-813e-8939c68e659b.png) 
I can add `introduced` and only those rows with `introduced` are selected (cfr. also ` Page: 46`,..) . When storing factors not explicitly (but as e.g. enum), you losse this capability.

Reply:
That seems to require both a header and quoted character variables. Maybe we should offer both options to the user: the verbose but large one and the cryptic but small one.

Reply:
BTW the current version of git2r requires 38 kB instead of 73kB to store the datafile from your example